cmd/pulumi-resource-docker-buildkit/main.go

```diff
@@ -231,54 +244,78 @@ func (k *dockerBuildkitProvider) dockerBuild(
 		return nil, err
 	}
 
+	var authToken string
 	if !username.IsNull() && !password.IsNull() {
-		cmd := exec.Command(
-			"docker", "login",
-			"-u", username.StringValue(), "--password-stdin",
-			registry["server"].StringValue(),
-		)
-		cmd.Stdin = strings.NewReader(password.StringValue())
-		if err := runCommand(ctx, k.host, urn, cmd); err != nil {
-			return nil, fmt.Errorf("docker login failed: %w", err)
+		auth := types.AuthConfig{
+			Username:      username.StringValue(),
+			Password:      password.StringValue(),
+			ServerAddress: registry["server"].StringValue(),
 		}
+		_, err := k.client.RegistryLogin(ctx, auth)
+		if err != nil {
+			return nil, fmt.Errorf("docker login failed: %v", err)
+		}
+		authConfigBytes, _ := json.Marshal(auth)
+		authToken = base64.URLEncoding.EncodeToString(authConfigBytes)
 	}
 
 	var platforms []string
 	for _, v := range inputs["platforms"].ArrayValue() {
 		platforms = append(platforms, v.StringValue())
 	}
-	cmd := exec.Command(
-		"docker", "buildx", "build",
-		"--platform", strings.Join(platforms, ","),
-		"--cache-from", name,
-		"--cache-to", "type=inline",
-		"-f", filepath.Join(context, dockerfile),
-		"-t", name, "--push",
-		context,
-	)
-	if err := runCommand(ctx, k.host, urn, cmd); err != nil {
-		return nil, fmt.Errorf("docker build failed: %w", err)
+	buildCtx, err := buildContext(context)
+	if err != nil {
+		return nil, fmt.Errorf("could not assemble docker build context for %#v: %v", context, err)
+	}
+	imageBuild := types.ImageBuildOptions{
+		Tags:       []string{name},
+		Dockerfile: dockerfile,
+		Context:    buildCtx,
+		Version:    types.BuilderBuildKit,
+		Remove:     true,
+		CacheFrom:  []string{name},
+		Platform:   strings.Join(platforms, ","),
+		// TODO: I don't know how to get --cache-to=type=internal set here /:
+	}
+	res, buildErr := k.client.ImageBuild(ctx, buildCtx, imageBuild)
+	if buildErr != nil {
+		return nil, fmt.Errorf("image build failed: %v", buildErr)
+	}
+	defer res.Body.Close()
+	if err = streamResponse(ctx, k.host, urn, diag.Info, diag.Error, res.Body); err != nil {
+		return nil, fmt.Errorf("could not stream build response: %v", err)
+	}
+
+	pushRes, pushErr := k.client.ImagePush(ctx, name, types.ImagePushOptions{
+		RegistryAuth: authToken,
+		Platform:     strings.Join(platforms, ","),
+	})
+	if pushErr != nil {
+		return nil, fmt.Errorf("image push failed: %v", pushErr)
+	}
+	if err = streamResponse(ctx, k.host, urn, diag.Info, diag.Error, pushRes); err != nil {
+		return nil, fmt.Errorf("could not stream push response: %v", err)
 	}
+	defer pushRes.Close()
```

Comment:
Shouldn't this go above the the line 296 `if` statement? If you error, it doesn't get closed.

Reply:
If you error, the result is nil, so closing isn't safe /: